integration/kotlinx-coroutines-play-services/src/Tasks.kt

```diff
@@ -105,3 +99,31 @@ public suspend fun <T> Task<T>.await(): T {
         }
     }
 }
+
+/**
+ * Awaits for completion of the created task without blocking a thread.
+ *
+ * Prefer this method over [Task.await] if a [Task] can be constructed with a [CancellationToken], to cancel the task
+ * if this function is cancelled.
+ *
+ * This suspending function is cancellable.
+ * If the [Job] of the current coroutine is cancelled or completed while this suspending function is waiting, this function
+ * stops waiting for the completion stage and immediately resumes with [CancellationException].
+ *
+ * @see [Task.await]
+ */
+@OptIn(ExperimentalContracts::class)
```

Comment:
The general idea "let's support bi-directional cancellation" looks good and worth supporting.

The proposed method though is quite verbose and non-idiomatic (submitted tests show it clearly), as well as hardly discoverable for newcomers.

I would like to keep things consistent and instead provide an overload of `await` with `CancellationToken`.
Also, then it's possible to simplify a code and use `CancellableContinuation.invokeOnCancellation` similarly to how it's done in other integrations, e.g. `Future.await()`

Reply:
I definitely agree that the extension on a lambda is pretty awkward, although I don't believe it will be possible to provide an overload of `Task.await` that takes a `CancellationToken` unless I'm missing something.

In cases like [`FusedLocationProviderClient.getCurrentLocation()`](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient#getCurrentLocation(int,%20com.google.android.gms.tasks.CancellationToken)) and [`FindAutocompletePredictionsRequest.Builder.setCancellationToken`](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient#public-tasklocation-getcurrentlocation-int-priority,-cancellationtoken-token), the `CancellationToken` is required in order to create the `Task`, so we need to setup the coroutine-managed `CancellationTokenSource` to give its `CancellationToken` to the caller.

I think the ideal solution would be to define a static extension on `Tasks` that takes the `block` to construct a `Task` with the `CancellationToken`, but that's blocked by https://youtrack.jetbrains.com/issue/KT-11968, so instead we could have a top-level function along the lines of `awaitTask` or `awaitConstructedTask`. I just pushed up a second commit with that alternative, which also changes `Task.await()` to call into `awaitTask`, since `Task.await` is the special case where the `Task` doesn't support being cancelled.

Let me know what you think!

Reply:
>the CancellationToken is required in order to create the Task, so we need to setup the coroutine-managed CancellationTokenSource to give its CancellationToken to the caller.

It can be created in advance, e.g.:
```
val token = CancellationTokenSource()
val task = createTask(token)
task.await(token)
```

The proposed solution has a significant restriction -- it confines the creation and cancalltion management into a single encapsulated function.
Meaning that if somebody _already_ has a cancellable task (and maybe invokes `.cancel()` manually from a few places) in their API, then it won't be possible for them to leverage cancellable await at all.

With the `await(token)`, on the other hand, it's easy to emulate the proposed API by introducing a small project-specific helper that doesn't even have to use any coroutines API beyond `await(task)`.

My vote is for providing a flexible and usable building block, rather than a specific-purpose non-extensible function.

Reply:
Ah, I see now, thank you for the insight! We can also do a very similar approach to support bi-directional cancellation for `asDeferred()` as well, via a passed in `CancellationTokenSource` which I think makes sense to tackle here as well.